src/librbd/mirror_snapshot/UnlinkPeerRequest.h

```diff
@@ -0,0 +1,84 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_LIBRBD_MIRROR_SNAPSHOT_UNLINK_PEER_REQUEST_H
+#define CEPH_LIBRBD_MIRROR_SNAPSHOT_UNLINK_PEER_REQUEST_H
+
+#include "include/buffer.h"
+
+#include <string>
+#include <set>
+
+struct Context;
+
+namespace librbd {
+
+struct ImageCtx;
+
+namespace mirror_snapshot {
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class UnlinkPeerRequest {
+public:
+  static UnlinkPeerRequest *create(ImageCtxT *image_ctx, uint64_t snap_id,
+                                   const std::string &mirror_peer_uuid,
+                                   Context *on_finish) {
+    return new UnlinkPeerRequest(image_ctx, snap_id, mirror_peer_uuid,
+                                 on_finish);
+  }
+
+  UnlinkPeerRequest(ImageCtxT *image_ctx, uint64_t snap_id,
+                    const std::string &mirror_peer_uuid, Context *on_finish)
+    : m_image_ctx(image_ctx), m_snap_id(snap_id),
+      m_mirror_peer_uuid(mirror_peer_uuid), m_on_finish(on_finish) {
+  }
+
+  void send();
+
+private:
+  /*
+   * @verbatim
+   *
+   * <start>
+   *    |
+   *    v
+   * REFRESH_IMAGE <------------------------\
+   *    |                                   |
+   *    | (peer found)                      |
+   *    |\-------------> UNLINK_PEER --> NOTIFY_UPDATE
```

Comment:
I worry about a potential failure mode where mirror snapshots are left behind after the last peer is unlinked. Perhaps the unlink should return an error if its the last peer so that it can just remove the snapshot w/o the need to unlink.